### ISSUE TITLE
Lock file updates to fuzzy ranges

### ISSUE THREAD

**Issue by bahmutov** [renovatebot/config-help#48]
First, we ❤️ Renovate app - it is just chugging along, keeping a few of our projects up to date painlessly. Perfect! But we have noticed a thing that is annoying.

We use `save-exact=true` in our `.npmrc` file, for example in https://github.com/cypress-io/schema-tools The updates from Renovate bot honor this and always update `package.json` to exact versions. But sometimes the `package-lock.json` is updated to _fuzzy_ version, for example https://github.com/cypress-io/schema-tools/commit/e5d7544f4c5a86ed6e6dc4ac1c78052abc7cd463 updated previously exact dependencies (transitive) to 

```diff
     "ts-node": {
-      "version": "6.0.3",
-      "resolved": "https://registry.npmjs.org/ts-node/-/ts-node-6.0.3.tgz",
-      "integrity": "sha512-ARaOMNFEPKg2ZuC1qJddFvHxHNFVckR0g9xLxMIoMqSSIkDc8iS4/LoV53EdDWWNq2FGwqcEf0bVVGJIWpsznw==",
+      "version": "6.0.4",
+      "resolved": "https://registry.npmjs.org/ts-node/-/ts-node-6.0.4.tgz",
+      "integrity": "sha512-20lw1igSKWxyb7G9OLoIn2duJWK9RfXqnT1WJVnGOhYmq6CK6QVqXCc7SSLBfqA/dEQG5jHfuJoYnA4byBa5Kw==",
       "dev": true,
       "requires": {
-        "arrify": "1.0.1",
-        "chalk": "2.4.1",
-        "diff": "3.5.0",
-        "make-error": "1.3.4",
-        "minimist": "1.2.0",
-        "mkdirp": "0.5.1",
-        "source-map-support": "0.5.6",
-        "yn": "2.0.0"
+        "arrify": "^1.0.0",
+        "chalk": "^2.3.0",
+        "diff": "^3.1.0",
+        "make-error": "^1.1.1",
+        "minimist": "^1.2.0",
+        "mkdirp": "^0.5.1",
+        "source-map-support": "^0.5.3",
+        "yn": "^2.0.0"
       },
```

Then when we pull down this commit and run `npm install` we usually have to commit and push `package-lock.json` back with exact dependencies - very annoying. For example this commit was only me pushing the changed lock file https://github.com/cypress-io/schema-tools/commit/46c205ee17bb7e0ef203f10df3a8142091b8dc45

Is this something one can configure? Why doesn't it honor the `save-exact` when updating the lock file?

**Comment by rarkins** [comment]
@bahmutov this is a `npm` bug I saw a couple of days ago and meant to follow up to see if it is affecting Renovate users. I may need to roll back to 5.x as a result. will update this once I've looked into it more.

**Comment by rarkins** [comment]
Found it: https://github.com/npm/npm/issues/20434

**Comment by bahmutov** [comment]
You are the best

**Comment by rarkins** [comment]
Actually it looks like not a bug - see my comment in the npm issue. Seems like a good refactor of lock file format that produces some diff noise in the meantime. Recording the actual dependency ranges that your direct (exact) dependencies use makes them much more traceable. Also, same as yarn does IIRC

**Comment by bahmutov** [comment]
But does it install the same dependencies tree then? Is the “requires” list only for description then? Hmm maybe then renovate should run instal and commit the lock file that results. I certainly don’t want to run npm install and see changes

**Comment by rarkins** [comment]
The ideal is that `npm install` doesn’t change lock files unless they need to change (ie lock file was invalid). `npm ci` sort of satisfies that. It nukes and recreates `node_modules` every time but is fast. 

Tried Renovate’s lock file maintenance?

**Comment by bahmutov** [comment]
Not sure I understand - renovate bot does update `package-lock.json` files, is there another setting? Also seems like this is difference between npm@5 and npm@6 where version 6 has less "wiggle" of dependencies.

**Comment by rarkins** [comment]
Doc: https://renovatebot.com/docs/configuration-options/#lockfilemaintenance

If you enable it, by default it will mean that once a week Renovate will completely regenerate your `package-lock.json` files from scratch to get the very latest versions of everything (i.e. all new patches and minor updates that fall within semver ranges). I mention it because if you had it enabled, then you would have had *all* packages updated to the npm@6 format by now, not just `ts-node`.

The lock file formats we discussed are backwards/forwards compatible and result in the same `node_modules`. What the new format does better is record what the transitive dependency *ranges* are. I wouldn't say there's any more or less "wiggle", just that the way the dependency trail is described/documented is different.

In summary, the new `package-lock.json` format is correct and intentional, so no configuration setting to change or fix. I recommend you get your team on npm@6 and consider running a fresh `npm install` for projects to get the "noise" over with, or perhaps enabling Renovate's "lock file maintenance" weekly to do it for you.

**Comment by bahmutov** [comment]
Thanks @rarkins - appreciate looking into this. We will switch to npm@6 and I will enable lock file maintenance too.
